### PR TITLE
[IMP] mail: channel_type cannot be changed

### PR DESCRIPTION
**Pull request by zel-odoo**
Get prevention when transforming the channel type.
There will be an UserError raised when the user tries to change the channel_type.

task-2774330

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/models/mail_channel.py [hunk 1]
```diff
@@ -239,6 +239,10 @@ def _unlink_except_all_employee_channel(self):
             raise UserError(_('You cannot delete those groups, as the Whole Company group is required by other modules.'))
 
     def write(self, vals):
+        if 'channel_type' in vals:
+            for channel in self:
+                if channel.channel_type:
+                    raise UserError(_('Cannot change channel type because \'%s\' is already been set to \'%s\'.') % (channel.name, channel.channel_type))
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:245]: Pending rebase on the other PR that was created following this issue, but there should be no condition on channel_type, it should always be set.

**Review by seb-odoo — COMMENTED** [review]
missing tag on the PR/commit title

File: addons/mail/models/mail_channel.py [hunk 2]
```diff
@@ -239,6 +239,10 @@ def _unlink_except_all_employee_channel(self):
             raise UserError(_('You cannot delete those groups, as the Whole Company group is required by other modules.'))
 
     def write(self, vals):
+        if 'channel_type' in vals:
+            for channel in self:
+                if channel.channel_type:
+                    raise UserError(_('Cannot change the channel type of \'%s\' because it is already been set.'), channel.name)
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:245]: I would apply the same principle TDE suggested you on the other PR: build the list of all the "failing" channels, and them list them in the error message.

Note: to actually detect change, you need to compare the given `channel_type` to the current channel type of each channel. It shouldn't prevent from giving to write the same value it already had.

File: addons/mail/models/mail_channel.py [hunk 3]
```diff
@@ -239,6 +239,10 @@ def _unlink_except_all_employee_channel(self):
             raise UserError(_('You cannot delete those groups, as the Whole Company group is required by other modules.'))
 
     def write(self, vals):
+        if 'channel_type' in vals:
+            failing_channels = self.sudo().filtered(lambda channel: channel.channel_type is not None and channel.channel_type != vals.get('channel_type'))
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:243]: We should merge PR about `channel_type` being required first, and remove condition on "is not None" from here.
Also "is not None" is probably not covering the expected cases anyway, as the ORM has a tendency to have `False` as as empty value, even for non-boolean fields.

Comment (zel-odoo) [reply addons/mail/models/mail_channel.py:243]: ok, and you can have a look at the 'required' task. it is ready to be reviewed!

File: addons/mail/tests/test_mail_channel.py [hunk 4]
```diff
@@ -448,19 +451,27 @@ def test_channel_unfollow_should_not_post_message_if_the_partner_has_been_remove
...
+        test_channel = self.env['mail.channel'].with_context(self._test_context).create({
+            'name': 'Channel',
+            'channel_type': 'channel',
+            'uuid': 'channel-uuid'
+        })
+        test_group = self.env['mail.channel'].with_context(self._test_context).create({
+            'name': 'Group',
+            'channel_type': 'group',
+            'uuid': 'group-uuid'
+        })
```
[the remark this answers is not in the document]
Comment (zel-odoo) [reply addons/mail/tests/test_mail_channel.py:463]: but then the uuid cannot be set. and this is important cause this is a test for the avatar.

File: addons/mail/tests/test_mail_channel.py [hunk 5]
```diff
@@ -448,19 +451,27 @@ def test_channel_unfollow_should_not_post_message_if_the_partner_has_been_remove
...
     def test_channel_should_generate_correct_default_avatar(self):
-        channel = self.env['mail.channel'].create({'name': '', 'uuid': 'test-uuid'})
-        bgcolor = html_escape('hsl(288, 51%, 45%)')  # depends on uuid
-        expceted_avatar_channel = (channel_avatar.replace('fill="#875a7b"', f'fill="{bgcolor}"')).encode()
-        expected_avatar_group = (group_avatar.replace('fill="#875a7b"', f'fill="{bgcolor}"')).encode()
+        test_channel = self.env['mail.channel'].with_context(self._test_context).create({
+            'name': 'Channel',
+            'channel_type': 'channel',
+            'uuid': 'channel-uuid'
+        })
```
[the remark this answers is not in the document]
Comment (zel-odoo) [reply addons/mail/tests/test_mail_channel.py:458]: same as `create_group()`

File: addons/mail/tests/test_mail_channel.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/tests/test_mail_channel.py:456]: But we could update the `uuid` of the created channel after the fact if needed?

[the remark this answers is not in the document]
Comment (zel-odoo) [reply addons/mail/tests/test_mail_channel.py:456]: indeed, I will do that.

**Review by seb-odoo — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**